compiler/optimizer/parallelize.go

```diff
@@ -2,189 +2,211 @@ package optimizer
 
 import (
 	"errors"
+	"fmt"
 
 	"github.com/brimdata/zed/compiler/ast/dag"
 	"github.com/brimdata/zed/order"
-	"github.com/brimdata/zed/pkg/field"
 	"golang.org/x/exp/slices"
 )
 
+// XXX Remove this and use native order.Direction in group-by.  See Issue #4505.
 func orderAsDirection(which order.Which) int {
 	if which == order.Asc {
 		return 1
 	}
 	return -1
 }
 
-// XXX assume the trunk is from a from op at seq.Ops[0] and we will
-// possible insert an operator at seq.Op[1]
-func (o *Optimizer) parallelizeTrunk(seq *dag.Sequential, trunk *dag.Trunk, replicas int) error {
-	from, ok := seq.Ops[0].(*dag.From)
-	if !ok {
-		return errors.New("internal error: parallelizeTrunk: entry is not a From")
+func (o *Optimizer) parallelizeScan(ops []dag.Op, replicas int) ([]dag.Op, error) {
+	// For now we parallelize only pool scans and no metadata scans.
+	// We can do the latter when we want to scale the performance of metadata.
+	if replicas < 2 {
+		panic(fmt.Sprintf("parallelizeScan: bad replicas: %d", replicas))
 	}
-	if len(from.Trunks) > 1 {
-		// No support for multi-trunk from's yet, which only arise for
-		// joins and other peculaliar mixins of different sources.
-		// We need to handle join parallelization differently, though
-		// the logic is not very far from what's here.
-		return nil
+	if len(ops) < 1 {
+		panic("parallelizeScan: short path")
 	}
-	if len(from.Trunks) == 0 {
-		return errors.New("internal error: no trunks in dag.From")
+	scan, ok := ops[0].(*dag.SeqScan)
+	if !ok {
+		return nil, errors.New("internal error: parallelizeScan: entry has no dag.SeqScan")
 	}
-	egressSortKey, err := o.sortKeyOfTrunk(trunk)
+	srcSortKey, err := o.sortKeyOfSource(scan)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	if len(egressSortKey.Keys) > 1 {
-		// XXX don't yet support multi-key ordering
-		return nil
+	if len(srcSortKey.Keys) > 1 {
+		// XXX Don't yet support multi-key ordering.  See Issue #2657.
+		return nil, nil
 	}
-	// This logic requires that there is only one trunk in the From,
-	// as checked above.
-	sortKey, err := o.sortKeyOfSource(trunk.Source, order.Nil)
-	if err != nil {
-		return err
+	if len(ops) == 1 && scan.Filter == nil {
+		return nil, nil
 	}
-	inlineSequentials(seq)
-	// Check that the path consisting of the original from
+	// concurrentPath will check that the path consisting of the original from
```

Comment:
Say "source" here since we no longer have dag.From?
```suggestion
	// concurrentPath will check that the path consisting of the original source
```